File: sdks/ruby/examples/basic_usage.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Basic usage example for the Scope Client Ruby SDK
#
# Before running, set your environment variables:
#   export SCOPE_ORG_ID=your_org_id
#   export SCOPE_API_KEY=your_api_key
#   export SCOPE_API_SECRET=your_api_secret
#   export SCOPE_API_URL=https://api.scope.io
#   export SCOPE_AUTH_API_URL=https://auth.scope.io
#
# Run with:
#   ruby examples/basic_usage.rb

require_relative '../lib/scope_client'

# Configure the client using credentials from environment
begin
  credentials = ScopeClient::Credentials::ApiKey.from_env
rescue ScopeClient::ConfigurationError => e
  puts "Configuration error: #{e.message}"
  puts 'Please set SCOPE_ORG_ID, SCOPE_API_KEY, SCOPE_API_SECRET, SCOPE_API_URL, and SCOPE_AUTH_API_URL'
  exit 1
end

ScopeClient.configure do |config|
  config.credentials = credentials
  config.cache_ttl = 300 # Cache prompts for 5 minutes
end

# Create a client instance
client = ScopeClient.client

puts '=== Scope Client Ruby SDK - Basic Usage ==='
puts

# Example 1: Get production version
puts '1. Getting production version...'
begin
  version = client.get_prompt_version('customer-greeting')
  puts "   Version: v#{version.version}"
  puts "   Status: #{version.status}"
  puts "   Variables: #{version.variables.join(', ')}"
  puts "   Content preview: #{version.content[0..100]}..."
rescue ScopeClient::NoProductionVersionError
  puts '   No production version exists'
rescue ScopeClient::NotFoundError
  puts '   Prompt not found'
end
puts

# Example 2: Render a prompt with variables
puts '2. Rendering a prompt...'
begin
  rendered = client.render_prompt('customer-greeting', {
                                    customer_name: 'Alice',
                                    company_name: 'Acme Corp',
                                    product: 'Widget Pro'
                                  })
  puts '   Rendered content:'
  puts "   #{rendered}"
rescue ScopeClient::MissingVariableError => e
  puts "   Missing variables: #{e.message}"
rescue ScopeClient::NotFoundError
  puts '   Prompt not found'
rescue ScopeClient::NoProductionVersionError
  puts '   No production version exists'
end
puts

# Example 3: Working with caching
puts '3. Caching demonstration...'
begin
  # First request - fetches from API
  start_time = Time.now
  client.get_prompt_version('customer-greeting')
  first_duration = Time.now - start_time
  puts "   First request: #{(first_duration * 1000).round(2)}ms"

  # Second request - served from cache
  start_time = Time.now
  client.get_prompt_version('customer-greeting')
  second_duration = Time.now - start_time
  puts "   Second request (cached): #{(second_duration * 1000).round(2)}ms"

  # Bypass cache
  start_time = Time.now
  client.get_prompt_version('customer-greeting', cache: false)
  third_duration = Time.now - start_time
  puts "   Third request (cache bypassed): #{(third_duration * 1000).round(2)}ms"
rescue ScopeClient::NotFoundError, ScopeClient::NoProductionVersionError
  puts '   Prompt not found or no production version'
end
puts

puts '=== Done ==='

File: sdks/python/examples/basic_usage.py
#!/usr/bin/env python3
"""Basic usage example for the Scope Client SDK.

This example demonstrates the basic functionality of the SDK including:
- Configuration
- Fetching prompt versions
- Rendering templates
- Using caching

Before running this example, set your environment variables:
    export SCOPE_ORG_ID="your-org-id"
    export SCOPE_API_KEY="your-api-key"
    export SCOPE_API_SECRET="your-api-secret"

Or modify the script to include your credentials directly.
"""

import scope_client
from scope_client import ApiKeyCredentials, Telemetry, RequestInfo, ResponseInfo


def main() -> None:
    # Configure the SDK using credentials from environment
    try:
        credentials = ApiKeyCredentials.from_env()
    except scope_client.ConfigurationError:
        print("Please set SCOPE_ORG_ID, SCOPE_API_KEY, and SCOPE_API_SECRET environment variables")
        return

    scope_client.configure(
        credentials=credentials,
        cache_enabled=True,
        cache_ttl=300,  # Cache for 5 minutes
    )

    # Optional: Set up telemetry to see requests
    def log_request(info: RequestInfo) -> None:
        print(f"[Request] {info.method} {info.url}")

    def log_response(info: ResponseInfo) -> None:
        print(f"[Response] {info.status_code} in {info.elapsed_ms:.2f}ms")

    Telemetry.on_request(log_request)
    Telemetry.on_response(log_response)

    # Create a client
    client = scope_client.client()

    print("\n=== Scope Client SDK Basic Usage ===\n")

    # Example 1: Get production version and render
    print("1. Getting production version...")
    try:
        version = client.get_prompt_version("greeting")
        print(f"   Content template: {version.content}")
        print(f"   Variables: {version.variables}")

        rendered = version.render(
            name="Alice",
            greeting="Hello",
        )
        print(f"   Rendered: {rendered}")
    except scope_client.NoProductionVersionError as e:
        print(f"   No production version for: {e.prompt_id}")
    except scope_client.NotFoundError:
        print("   Prompt 'greeting' not found")
    print()

    # Example 2: Using the render_prompt convenience method
    print("2. Using render_prompt method...")
    try:
        rendered = client.render_prompt(
            "greeting",
            {"name": "Bob", "greeting": "Hi"},
            label="latest",
        )
        print(f"   Rendered: {rendered}")
    except (scope_client.NotFoundError, scope_client.NoProductionVersionError):
        print("   Prompt not found or no version available")
    print()

    # Example 3: Cache demonstration
    print("3. Cache demonstration...")
    print("   First request (cache miss):")
    try:
        client.get_prompt_version("greeting")
    except (scope_client.NotFoundError, scope_client.NoProductionVersionError):
        pass

    print("   Second request (cache hit - no HTTP request):")
    try:
        client.get_prompt_version("greeting")
    except (scope_client.NotFoundError, scope_client.NoProductionVersionError):
        pass

    print("   Request with cache bypass:")
    try:
        client.get_prompt_version("greeting", cache=False)
    except (scope_client.NotFoundError, scope_client.NoProductionVersionError):
        pass
    print()

    # Example 4: Clear cache
    print("4. Clearing cache...")
    client.clear_cache()
    print("   Cache cleared")

    # Clean up
    client.close()
    print("\n=== Done ===")


if __name__ == "__main__":
    main()
